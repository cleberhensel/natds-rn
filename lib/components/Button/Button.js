/* eslint-disable complexity */
import React from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

import { ButtonWrapper, ButtonText, IconPress } from './Button.styles';
import { loadIcon } from '../../utils';
import { ThemeNaturaLight } from '../../themes/ThemeNaturaLight';
import { NatSpace } from '../NatSpace';

const styles = {
  elevation: 1,
  shadowColor: '#000',
  shadowOffset: {
    height: 2,
    width: 0,
  },
  shadowOpacity: 0.24,
};

export const Button = ({
  color,
  iconLeft,
  iconRight,
  fixedBottom,
  label,
  node,
  noUpperCase,
  onPress,
  onPressLeft,
  onPressRight,
  showIconLeft,
  showIconRight,
  theme,
  type,
}) => {
  const renderIcon = (iconData, iconPress, position) => {
    const iconAttributes = {
      color: ThemeNaturaLight.palette.text.primary,
      size: 24,
    };

    let styleCustom = {};

    if (position === 'right') {
      styleCustom = { position: 'absolute', right: 0 };
    } else {
      styleCustom = { left: 0, position: 'absolute' };
    }

    return (
      <View style={styleCustom}>
        <IconPress onPress={iconPress} position={position}>
          { loadIcon(iconData, iconAttributes) }
        </IconPress>
      </View>
    );
  };

  const renderTitle = () => {
    if (label !== '') {
      return (
        <ButtonText
            type={type}
            theme={theme}
            color={color}
            accessibilityLabel={`${label}`}
          >
            {noUpperCase ? label : label.toUpperCase()}
          </ButtonText>
      );
    }

    return node;
  };

  return (
    <NatSpace
      backgroundColor={fixedBottom ? theme.palette.background.paper : 'transparent'}
      marginBottom="none"
      marginLeft="none"
      marginRight="none"
      marginTop="none"
      paddingBottom={fixedBottom ? 'tiny' : 'none'}
      paddingLeft={fixedBottom ? 'small' : 'none'}
      paddingRight={fixedBottom ? 'small' : 'none'}
      paddingTop={fixedBottom ? 'small' : 'none'}
    >
      <ButtonWrapper
        accessibilityRole="button"
        accessible
        color={color}
        disabled={color === 'disabled'}
        onPress={onPress}
        style={(type === 'primary' && color !== 'disabled') && styles}
        theme={theme}
        type={type}
      >
        { (showIconLeft && iconLeft !== null) && renderIcon(iconLeft, onPressLeft, 'left') }
        { renderTitle() }
        { showIconRight && renderIcon(iconRight, onPressRight, 'right') }
      </ButtonWrapper>
    </NatSpace>
  );
};

Button.propTypes = {
  color: PropTypes.string,
  fixedBottom: PropTypes.bool,
  iconLeft: PropTypes.shape(),
  iconRight: PropTypes.shape(),
  label: PropTypes.string,
  node: PropTypes.node,
  noUpperCase: PropTypes.bool,
  onPress: PropTypes.func.isRequired,
  onPressLeft: PropTypes.func,
  onPressRight: PropTypes.func,
  showIconLeft: PropTypes.bool,
  showIconRight: PropTypes.bool,
  state: PropTypes.string,
  theme: PropTypes.shape(),
  type: PropTypes.string,

};

Button.defaultProps = {
  color: 'default',
  fixedBottom: false,
  iconLeft: null,
  iconRight: null,
  label: '',
  node: null,
  noUpperCase: false,
  onPressLeft: () => {},
  onPressRight: () => {},
  showIconLeft: true,
  showIconRight: true,
  theme: ThemeNaturaLight,
  type: 'primary',
};
