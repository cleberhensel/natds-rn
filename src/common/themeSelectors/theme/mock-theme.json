{
  "avatarSizes": {
    "huge": {
      "fontSize": "1.5rem",
      "size": 80
    },
    "large": {
      "fontSize": "1rem",
      "size": 48
    },
    "small": {
      "fontSize": ".875rem",
      "size": 32
    },
    "standard": {
      "fontSize": "1rem",
      "size": 40
    },
    "tiny": {
      "fontSize": ".625rem",
      "size": 24
    }
  },
  "buttonSizes": {
    "large": {
      "minHeight": 48,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16
    },
    "medium": {
      "minHeight": 40,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 12
    },
    "small": {
      "minHeight": 32,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8
    }
  },
  "colorTokens": {
    "colorAlert": "#E74627",
    "colorBackground": "#FAFAFA",
    "colorHighEmphasis": "#333333",
    "colorHighlight": "#000000",
    "colorLink": "#227BBD",
    "colorLowEmphasis": "#BBBBBB",
    "colorMediumEmphasis": "#777777",
    "colorOnAlert": "#FFFFFF",
    "colorOnBackground": "#333333",
    "colorOnLink": "#FFFFFF",
    "colorOnPrimary": "#333333",
    "colorOnPrimaryDark": "#333333",
    "colorOnPrimaryLight": "#333333",
    "colorOnSecondary": "#000000",
    "colorOnSecondaryDark": "#000000",
    "colorOnSecondaryLight": "#000000",
    "colorOnSuccess": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorOnWarning": "#333333",
    "colorPrimary": "#F4AB34",
    "colorPrimaryDark": "#EF8426",
    "colorPrimaryLight": "#FEFDE8",
    "colorSecondary": "#FF6B0B",
    "colorSecondaryDark": "#FF5808",
    "colorSecondaryLight": "#FFF8E1",
    "colorSuccess": "#569A32",
    "colorSurface": "#FFFFFF",
    "colorWarning": "#FCC433"
  },
  "elevation": {
    "0": "none",
    "1": "0 1px 1px 0 rgba(0, 0, 0, 0.14)",
    "12": "0 12px 17px 0 rgba(0, 0, 0, 0.14)",
    "16": "0 16px 24px 0 rgba(0, 0, 0, 0.14)",
    "2": "0 2px 2px 0 rgba(0, 0, 0, 0.14)",
    "24": "0 24px 38px 0 rgba(0, 0, 0, 0.14)",
    "3": "0 3px 4px 0 rgba(0, 0, 0, 0.14)",
    "4": "0 4px 5px 0 rgba(0, 0, 0, 0.14)",
    "6": "0 6px 10px 0 rgba(0, 0, 0, 0.14)",
    "8": "0 8px 10px 0 rgba(0, 0, 0, 0.14)",
    "9": "0 9px 12px 0 rgba(0, 0, 0, 0.14)"
  },
  "fontSize": {
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    }
  },
  "iconSizes": {
    "huge": 144,
    "hugex": 192,
    "hugexx": 256,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "medium": 64,
    "micro": 16,
    "small": 32,
    "standard": 48,
    "tiny": 24
  },
  "opacity": {
    "00": 0,
    "04": 0.04,
    "08": 0.08,
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "circle": "50%",
    "large": 8,
    "medium": 4,
    "none": 0,
    "small": 2
  },
  "sizes": {
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "medium": 48,
    "mediumx": 56,
    "micro": 4,
    "none": 0,
    "semi": 32,
    "semix": 40,
    "small": 16,
    "standard": 24,
    "tiny": 8,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingHuge": 64,
    "spacingLarge": 48,
    "spacingMicro": 4,
    "spacingNone": 0,
    "spacingSemi": 32,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingTiny": 8
  }
}
