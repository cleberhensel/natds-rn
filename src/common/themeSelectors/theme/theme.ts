import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutTokens = Omit<TokensWithoutColor, 'colorTokens'>
export type Theme = TokensWithoutTokens & { colorTokens: IColorThemeTokens }

type ThemeMode = 'dark' | 'light';

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
