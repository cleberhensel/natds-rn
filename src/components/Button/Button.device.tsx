import { storiesOf } from '@storybook/react-native';
import {
  contained, outlined, text, all,
} from './Button.stories';

storiesOf('Button', module)
  .add('all', all)
  .add('contained', contained)
  .add('outlined', outlined)
  .add('text', text);
