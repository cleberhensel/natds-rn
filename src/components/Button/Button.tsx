import React from 'react';
import styled, { withTheme } from 'styled-components/native';
import {
  getTheme,
  getColorPrimary,
  getColorOnPrimary,
  getColorHighEmphasis,
  Theme,
  getButtonPropsBySize,
  getRadiusPropsBySize,
  getShadowBySize,
} from '../../common/themeSelectors';

export type ButtonTypes = 'contained' | 'outlined' | 'text'

export interface ButtonProps {
  /**
   * The button content
   */
  text: string
  type?: ButtonTypes
  theme: Theme,
  testID?: string
}

interface ButonBase {
  type: ButtonTypes
  theme: Theme
}

const getButtonStyles = (theme: Theme) => ({
  contained: {
    background: getColorPrimary(theme),
  },
  outlined: {
    borderColor: getColorPrimary(theme),
    borderWidth: 1,
  },
});

const ButtonBase = styled.TouchableOpacity<ButonBase>(({ type, theme }) => ({
  borderRadius: getRadiusPropsBySize(theme, 'medium'),
  ...getButtonStyles(theme)[type],
  ...getButtonPropsBySize(theme, 'medium'),
}));

const Text = styled.Text<ButonBase>`
color: ${
  props => (
    props.type === 'contained'
      ? getTheme(getColorOnPrimary)
      : getTheme(getColorHighEmphasis))
};
`;

const getShadowByType = (type: ButtonTypes, theme: Theme) => {
  if (type === 'contained') {
    return getShadowBySize(theme, '2');
  }

  return {};
};

const ButtonComponent = ({
  theme, text, type = 'contained', testID = 'button',
}: ButtonProps) => (
  <ButtonBase
    testID={testID}
    style={getShadowByType(type, theme)}
    type={type}
  >
    <Text type={type}>{text}</Text>
  </ButtonBase>
);

export const Button = withTheme(ButtonComponent);
