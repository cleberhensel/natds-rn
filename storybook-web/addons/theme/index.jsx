import React, { useEffect, Fragment } from 'react';
import { useAddonState, Consumer } from '@storybook/api';
import { TabButton, WithTooltip, TooltipLinkList } from '@storybook/components';

import { PANEL_ID, CHANGE, PARAM_KEY } from './shared';
import { STORY_CHANGED } from '@storybook/core-events';

import './styles.css';

const DEFAULT_THEME = {
  name: 'natura',
  type: 'light'
};

function parseTheme(themes) {
  return Object.entries(themes).reduce((accum, [name, variants]) => {
    Object.entries(variants).forEach(function ([type]) {
      accum.push({ name, type })
    })

    return accum;
  }, []);
}

const getDisplayedItems = (list, onChange, selectedItem) => {
  if (!list.length) {
    return [];
  }

  return [
    ...list.map((params) =>
      createBackgroundSelectorItem(params, onChange, selectedItem)
    ),
  ];
};

const createBackgroundSelectorItem = ({ name, type }, onChange, selectedItem) => ({
  id: `${name} - ${type}`,
  title: `${name} - ${type}`,
  active: selectedItem.name === name && selectedItem.type === type,
  onClick: () => onChange({ name, type })
});

const mapper = ({ api, state }) => {
  const story = state.storiesHash[state.storyId];
  const { themes = {} } = story && api.getParameters(story.id, PARAM_KEY) || {};

  return { items: parseTheme(themes) };
};

export default function Theme(props) {
  const { channel, api } = props;
  const [currentTheme, changeTheme] = useAddonState(PANEL_ID, DEFAULT_THEME);
  const { disabled } = props.api.getCurrentParameter(PARAM_KEY) || {};

  const handleChange = (params) => {
    changeTheme(params);
    channel.emit(CHANGE, params);
  };

  useEffect(() => {
    channel.on(STORY_CHANGED, () => {
      changeTheme(DEFAULT_THEME);
    });

    return () => (channel.removeListener(STORY_CHANGED))
  }, []);

  if (disabled) return null;

  return (
    <Consumer filter={mapper}>
      {({ items }) => {
        return items.length ? (
          <Fragment>
            <WithTooltip
              placement="top"
              trigger="click"
              closeOnClick
              tooltip={({ onHide }) => (
                <TooltipLinkList
                  links={getDisplayedItems(items, i => {
                    handleChange(i);
                    onHide();
                  }, currentTheme)}
                />
              )}
            >
              <TabButton
                key="Theme"
                title="Change the Theme">
                {currentTheme.name} - {currentTheme.type}
              </TabButton>
            </WithTooltip>
          </Fragment>
        ) : null;
      }}
    </Consumer>
  );
}
