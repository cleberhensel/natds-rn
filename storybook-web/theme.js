import { create } from '@storybook/theming/create';

export default create({
  appBg: '#333',
  appBorderRadius: 0,
  appContentBg: '#fff',
  barBg: '#333',
  barSelectedColor: '#F49B00',
  barTextColor: '#f0f0f0',
  base: 'light',
  brandImage: '/logo-natura-negative-vertical.svg',
  brandTitle: 'natds-js',
  brandUrl: '/',
  colorPrimary: '#F49B00',
  colorSecondary: '#EF8426',
  textColor: '#333',
  textInverseColor: '#fff',
});
