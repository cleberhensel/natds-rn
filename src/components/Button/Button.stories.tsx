
import React from 'react';
import { ContainerRow } from '../../common/HelperComponents/ThemeHelper.styles';
import { Button } from './Button';

export default {
  component: Button,
  parameters: {
    componentSubtitle: 'A usefull button component.',
  },
  title: 'Components|Button',
};

const defaultProps = ({
  onPress: () => {},
  text: 'DEFAULT',
});

export const contained = () => (
  <ContainerRow style={{ padding: 30 }} >
    <Button { ...defaultProps }/>
  </ContainerRow>
);

export const outlined = () => (
  <ContainerRow >
    <Button type="outlined" { ...defaultProps } />
  </ContainerRow>
);

export const text = () => (
  <ContainerRow >
    <Button type="text" { ...defaultProps } />
  </ContainerRow>
);
