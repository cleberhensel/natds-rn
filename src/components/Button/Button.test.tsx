import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import '@testing-library/jest-native/extend-expect';
import { ThemeProvider } from 'styled-components/native';
import { Button } from './Button';
import theme from '../../common/themeSelectors/theme/mock-theme.json';

jest.mock('../../common/themeSelectors', () => (
  {
    getButtonPropsBySize: () => ({ background: '#AEAEAE' }),
    getColorHighEmphasis: () => '#FAF3E3',
    getColorOnPrimary: () => '#F4F4',
    getColorPrimary: () => '#FFFFFF',
    getRadiusPropsBySize: () => 42,
    getShadowBySize: () => ({ shadowColor: '#AEAEAE' }),
    getTheme: fn => fn(),
  }));

describe('Button component', () => {
  it('Should render button with a text', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button text="texto legal"/>
      </ThemeProvider>,
    );

    expect(queryByTestId('button')).toHaveTextContent('texto legal');
  });

  it('Should render button with default props', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button text="texto legal"/>
      </ThemeProvider>,
    );

    expect(queryByTestId('button')?.props).toHaveProperty('type', 'contained');
  });

  it('Should render button with type props', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button type="outlined" text="texto legal"/>
      </ThemeProvider>,
    );

    expect(queryByTestId('button')).toHaveStyle({
      borderWidth: 1,
    });
  });

  it('Should render button component outlined', () => {
    const button = renderer.create(
      <ThemeProvider theme={theme}>
        <Button type="outlined" text="texto legal"/>
      </ThemeProvider>,
    ).toJSON();

    expect(button).toMatchSnapshot();
  });

  it('Should render button component contained', () => {
    const button = renderer.create(
      <ThemeProvider theme={theme}>
        <Button text="texto legal"/>
      </ThemeProvider>,
    ).toJSON();

    expect(button).toMatchSnapshot();
  });

  it('Should render button component text', () => {
    const button = renderer.create(
      <ThemeProvider theme={theme}>
        <Button type="text" text="texto legal"/>
      </ThemeProvider>,
    ).toJSON();

    expect(button).toMatchSnapshot();
  });
});
