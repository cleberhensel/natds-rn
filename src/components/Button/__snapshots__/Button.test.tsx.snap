// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Should render button component contained 1`] = `
<
  style={
    Array [
      Object {
        "backgroundColor": "#AEAEAE",
        "borderRadius": 42,
      },
      Object {
        "shadowColor": "#AEAEAE",
      },
    ]
  }
  testID="button"
  type="contained"
>
  <Text
    style={
      Array [
        Object {
          "color": "#F4F4",
        },
      ]
    }
    type="contained"
  >
    texto legal
  </Text>
</>
`;

exports[`Button component Should render button component outlined 1`] = `
<
  style={
    Array [
      Object {
        "backgroundColor": "#AEAEAE",
        "borderColor": "#FFFFFF",
        "borderRadius": 42,
        "borderWidth": 1,
      },
      Object {},
    ]
  }
  testID="button"
  type="outlined"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FAF3E3",
        },
      ]
    }
    type="outlined"
  >
    texto legal
  </Text>
</>
`;

exports[`Button component Should render button component text 1`] = `
<
  style={
    Array [
      Object {
        "backgroundColor": "#AEAEAE",
        "borderRadius": 42,
      },
      Object {},
    ]
  }
  testID="button"
  type="text"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FAF3E3",
        },
      ]
    }
    type="text"
  >
    texto legal
  </Text>
</>
`;
